scipy/optimize/_remove_redundancy.py

```diff
@@ -230,6 +231,37 @@ def _remove_redundancy_dense(A, rhs):
     return A_orig[keep, :], rhs[keep], status, message
 
 
+class LUSolver:
+    def __init__(self, m):
+        I = eye(m, format='csc')
+        self.fl = splu(I, permc_spec="NATURAL")
+        self.fu = splu(I, permc_spec="NATURAL")
+        self.L = eye(m, format='csc')
+        self.U = eye(m, format='csc')
+
+    def solve(self, b, trans=False):
+        if trans:
+            y = self.fu.solve(b, 'T')
+            x = self.fl.solve(y, 'T')
+        else:
+            y = self.fl.solve(b)
+            x = self.fu.solve(y)
+        return x
+
+    def update(self, v, j):
+        L = self.L
+        U = self.U
+        m, n = L.shape
+
+        u = self.fl.solve(v.toarray())
+        U[:j+1, j] = u[:j+1]
```

Comment:
`U[:j+1, j] = u[:j+1]` and `L[j+1:, j] += (l/piv)` are the bottlenecks because updating the sparsity structure of a CSC matrix is slow. I tried making `self.L` and `self.U` LIL and DOK matrices, but SuperLU needs to convert them to CSC, so it doesn't save any time. Any way around this?

Reply:
One way I can think to speed up the update of U is to  directly update U.data, U.indptr, U.indices. Not the most readable solution but I suspect it will be one of the fastest solutions. 
Is the array u[:j+1] expected to be all nonzeros. i.e. would U end up being a dense matrix?

Reply:
Yeah I was actually coming back here to write that myself. I think I'll subclass `csc_matrix` and write my own update routine. Besides the nonzeros being added, no more than `n` entries need to be moved each update, and they're all ones. I can double the size of the underlying array each time it runs out of room to keep the amortized computational complexity of copying low.

Reply:
I don't know how dense U will end up being, but this algorithm was written with sparsity in mind. Supposedly the basis matrix can be refactored from scratch every so often to preserve sparsity. I anticipate difficulties with that, because the paper is not explicit about how to deal with the permutation matrices, but we'll see.

Reply:
This is indeed much faster.
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   299                                               @profile
   300                                               def update(self, v, j):
   301       742      38585.0     52.0      0.0          L = self.L
   302       742      22328.0     30.1      0.0          U2 = self.U2
   303       742      14936.0     20.1      0.0          U = self.U
   304       742      36694.0     49.5      0.0          m, n = L.shape
   305
   306       742   11532889.0  15543.0      1.8          u = self.fl.solve(v.toarray())
   307       742    1150357.0   1550.3      0.2          U.update(j, u, True)
   308       742   62228490.0  83865.9      9.7          U2[:j+1, j] = u[:j+1]
```
`U2` is the `csc_matrix`, `U` is the subclass.

Reply:
This helped, but for larger problems stuffing the already factorized matrix into SuperLU objects in order to perform triangular solves becomes a bottleneck. I'll see if a custom triangular solve routine helps.

Reply:
A custum triangular solve routine doesn't help. Timing the custom triangular solve next to SuperLU's solve:


I don't think the solve routine is written poorly.
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   276                                               @profile
   277                                               def solve(self, b, trans=False):
   278      1291      46978.0     36.4      0.0          n = self.shape[0]
   279      1291      19692.0     15.3      0.0          if not trans:
   280      1291      98040.0     75.9      0.0              x = np.zeros(self.shape[0])
   281                                           #            b = b.reshape(-1,1)
   282      1291      35077.0     27.2      0.0              b = b.ravel()
   283      1291      75259.0     58.3      0.0              p = np.zeros(self.shape[0])
   284      1291      21339.0     16.5      0.0              if self.upper:
   285                                                           cols = range(-1, -len(x)-1, -1)
   286                                                       else:
   287      1291      44748.0     34.7      0.0                  cols = range(len(x))
   288   1667972   23894092.0     14.3      3.5              for i in cols:
   289                                                           # col = self[:,i].todense()
   290   1666681   87984804.0     52.8     12.9                  col = np.zeros(n)
   291                                           #                col_sparse = self[:,i]  # too slow
   292   1666681   38762053.0     23.3      5.7                  j1 = self.indptr[i]
   293   1666681   35457509.0     21.3      5.2                  j2 = self.indptr[i+1]
   294   1666681   41360170.0     24.8      6.1                  inz = self.indices[j1:j2]
   295   1666681  122782028.0     73.7     18.1                  col[inz] = self.data[j1:j2]
   296   1666681  120402197.0     72.2     17.7                  x[i] = (b-p)[i]/col[i]
   297   1666681  208870871.0    125.3     30.7                  p += x[i]*col
   298      1291      57105.0     44.2      0.0              return x.ravel()
```
I could probably reduce the number of executions by 1/2 by taking advantage of the `my_csc_matrix` object's knowledge of the number of non-identify columns it has, but this wouldn't fix the problem. Each operation is nearly as fast as a simple variable assignment - there are just too many of them. I could try Cythonizing it, but before I go that route I'll look into SuperLU a bit deeper to see if there's a better way of creating the SuperLU objects from the existing factorss.

Reply:
I have just looked superficially, but it seems that one can access the umfpack L and U matrices directly. Perhaps we can only have this optimization if umfpack is available. 
https://github.com/scikit-umfpack/scikit-umfpack/blob/master/scikits/umfpack/interface.py

Reply:
And that would still be worth it, but it's unclear whether those attributes are writable, and I can't test it on this computer. For SuperLU, you can read them, but not write to them:
```
ipdb> self.fl
<SuperLU object at 0x0000024315E50780>

ipdb> self.fl.L
<8x8 sparse matrix of type '<class 'numpy.float64'>'
        with 8 stored elements in Compressed Sparse Column format>

ipdb> self.fl.L = self.fl.L
*** AttributeError: attribute 'L' of 'SuperLU' objects is not writable
```

Reply:
Also, SuperLU has an option:
```
Fact
Specifies whether or not the factored form of the matrix A is supplied on entry, and if not,
how the matrix A will be factorized base on the previous history, such as factor from scratch,
reuse Pc and/or Pr, or reuse the data structures of L and U. fact can be one of:
– DOFACT: the matrix A will be factorized from scratch.
...
– FACTORED: the factored form of A is input
```
And the SciPy interface allows me to specify these options, but only `DOFACT` works for me. The rest crash or hang without error.

Reply:
> For SuperLU, you can read them, but not write to them

Same with scikit-umfpack.